crates/astria-sequencer/src/accounts/ics20_transfer.rs

```diff
@@ -0,0 +1,334 @@
+use anyhow::{
+    Context,
+    Result,
+};
+use ibc_types::{
+    core::channel::{
+        channel::Order as ChannelOrder,
+        msgs::{
+            MsgAcknowledgement,
+            MsgChannelCloseConfirm,
+            MsgChannelCloseInit,
+            MsgChannelOpenAck,
+            MsgChannelOpenConfirm,
+            MsgChannelOpenInit,
+            MsgChannelOpenTry,
+            MsgRecvPacket,
+            MsgTimeout,
+        },
+        ChannelId,
+        PortId,
+    },
+    transfer::acknowledgement::TokenTransferAcknowledgement,
+};
+use penumbra_ibc::component::app_handler::{
+    AppHandler,
+    AppHandlerCheck,
+    AppHandlerExecute,
+};
+use penumbra_proto::penumbra::core::component::ibc::v1alpha1::FungibleTokenPacketData;
+use penumbra_storage::{
+    StateRead,
+    StateWrite,
+};
+use proto::native::sequencer::v1alpha1::{
+    asset::{
+        IbcAsset,
+        Id,
+    },
+    Address,
+};
+
+use super::state_ext::StateWriteExt;
+use crate::accounts::state_ext::StateReadExt;
+
+/// The ICS20 transfer handler.
+/// See [here](https://github.com/cosmos/ibc/blob/main/spec/app/ics-020-fungible-token-transfer/README.md)
+/// for the specification which this is based on.
+#[derive(Clone)]
+pub(crate) struct Ics20Transfer {}
+
+#[async_trait::async_trait]
+impl AppHandlerCheck for Ics20Transfer {
+    async fn chan_open_init_check<S: StateRead>(_: S, msg: &MsgChannelOpenInit) -> Result<()> {
+        if msg.ordering != ChannelOrder::Unordered {
+            anyhow::bail!("channel order must be unordered for Ics20 transfer");
+        }
+
+        if msg.version_proposal != "ics20-1".to_string().into() {
```

Comment:
Unnecessary allocation:
```suggestion
        if msg.version_proposal.as_str() != "ics20-1" {
```